src/main/java/net/minecraftforge/client/model/generators/BlockModelBuilder.java

```diff
@@ -18,7 +26,231 @@
+         */
+        public RootTransformBuilder rightRotation(Quaternion rightRotation)
+        {
+            Preconditions.checkNotNull(rightRotation, "Rotation must not be null");
+            this.rightRotation = rightRotation;
+            return this;
+        }
+
+        /**
+         * Sets the scale of the root transform.
+         *
+         * @param scale the scale
+         * @return this builder
+         * @throws NullPointerException if {@code scale} is {@code null}
+         */
+        public RootTransformBuilder scale(float scale)
+        {
+            return scale(new Vector3f(scale, scale, scale));
+        }
+
+        /**
+         * Sets the scale of the root transform.
+         *
+         * @param scale the scale vector
+         * @return this builder
+         * @throws NullPointerException if {@code scale} is {@code null}
+         */
+        public RootTransformBuilder scale(Vector3f scale)
+        {
+            Preconditions.checkNotNull(scale, "Scale must not be null");
+            this.scale = scale;
+            return this;
+        }
+
+        /**
+         * Sets the root transform.
+         *
+         * @param transformation the transformation to use
+         * @return this builder
+         * @throws NullPointerException if {@code transformation} is {@code null}
+         */
+        public RootTransformBuilder transform(Transformation transformation)
+        {
+            Preconditions.checkNotNull(transformation, "Transformation must not be null");
+            this.translation = transformation.getTranslation();
+            this.leftRotation = transformation.getLeftRotation();
+            this.rightRotation = transformation.getRightRotation();
+            this.scale = transformation.getScale();
+            return this;
+        }
+
+        /**
+         * Sets the origi of the root transform.
+         *
+         * @param origin the origin vector
+         * @return this builder
+         * @throws NullPointerException if {@code origin} is {@code null}
+         */
+        public RootTransformBuilder origin(Vector3f origin)
+        {
+            Preconditions.checkNotNull(origin, "Origin must not be null");
+            this.origin = origin;
+            this.originName = null;
+            return this;
+        }
+
+        /**
+         * Sets the origin of the root transform.
+         *
+         * @param origin the origin name
+         * @return this builder
+         * @throws NullPointerException if {@code origin} is {@code null}
+         * @throws IllegalArgumentException if {@code origin} is not {@code center}, {@code corner} or {@code opposing-corner}
+         */
+        public RootTransformBuilder origin(String origin)
+        {
+            Preconditions.checkNotNull(origin, "Origin must not be null");
+            Preconditions.checkArgument(origin.equals("center") || origin.equals("corner") || origin.equals("opposing-corner"), "Invalid value for origin name.");
```

Comment:
We should be using an enum for this instead of requiring hardcoded strings. The enum should probably be `StringRepresentable`, with the enum values should include the friendly name (these three strings) and the origin vector as seen in the fields of `TransformationHelper.Deserializer`. 

Accordingly, `TransformationHelper.Deserializer#parseOrigin` would be modified to use the names from the enum values rather than hardcoded strings.

Reply:
An enum does sound like a good idea, is it a good idea to place the enum within TransformationHelper with the rest of the transformation logic, inside that common.util package or in the model package somewhere?

Reply:
Having it either be a nested class of `TransformationHelper` or a nested class of `BlockModelBuilder#RootTransformBuilder` should be fine. The former would be a more logical grouping of transformation-related classes, but the latter would be better in terms of being close to where the enumeration is actually used. I'd probably say the latter, but your choice between the two.

Reply:
In 66ff28c6a3c1cfc610b99932ebe43c937e5d6089 I have added an enum to RootTransformBuilder, let me know if you'd like anything changed with it, I've implemented it as best as I could think :)